themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md

```diff
@@ -0,0 +1,87 @@
+
+* The code handling the Kubernetes cluster could reference the VPC ID and subnet IDs from the base infrastructure layer.
+* Similarly, the code that was responsible for deploying the online store services could reference the necessary Kubernetes connection details for the Kubernetes cluster.
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Fortunately, adding stack outputs after the fact is easy and generally has no impact whatsoever on existing infrastructure.
+
+Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.
+
+While stack references are generally straightforward, there are some recommended practices to observe around the use of stack outputs and stack references:
+
+1. With regard to stack outputs, **only export what is needed.** If there's a need for it to be accessed from outside the stack, export it; otherwise, don't. It's easy to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're only modifying the stack object itself).
+2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.
+3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in only a single call to the backend. Be aware, however, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of additional work required in this situation.
+4. In line with parameterizing as much of your code as possible, **also be sure to parameterize your stack references.** A stack reference is built using an organization name, a project name, and a stack name. Don't hardcode these values; instead, use configuration values to allow the users to specify from which source stack(s) the values will be referenced.
```

Comment:
I can't think of a use case where organization would need to be templates, but stack name definitely should be. It might be nice to show a default value that is the current stack name, e.g.

```
const stack = new pulumi.Config().getStack().split("/")[2];
const stackRef = new pulumi.StackReference(`my-organization/workshop-aws-proton-base-infra/${stack}`);
```

Double check the above code in an actual program before publishing - I can't vouch for its correctness, but it's something like that.

Reply:
To me, it seems entirely plausible that a company could have two organizations, with users that are members of both organizations, and there be a need to read information across the organizational boundary via a stack reference (two different business units, perhaps?).

Reply:
+1 to add some code here to show what "parameterize your stack references" looks like.

Reply:
Code added in 17aacd6f, along with references to the code in other parts of the post